File: Difficulty: Medium/Longest Subarray with Sum K/longest-subarray-with-sum-k.cpp
class Solution {
  public:
    int longestSubarray(vector<int>& arr, int k) {
        // code here
        unordered_map<int,int>mp;
        int maxlen = INT_MIN;
        int csum = 0;
        for(int i = 0; i < arr.size(); i++){
            csum += arr[i];
            
            if(csum == k) maxlen = i + 1;
            
            if(mp.find(csum - k) != mp.end()){
                maxlen = max(maxlen, i - mp[csum - k]);
            }
            
            if(mp.find(csum) == mp.end()){
                mp[csum] = i;
            }
        }
        return maxlen == INT_MIN ? 0 : maxlen;
    }
};

File: Difficulty: Medium/Longest Subarray with Sum K/longest-subarray-with-sum-k.js
/**
 * @param {number[]} arr
 * @param {number} k
 * @returns {number}
 */
class Solution {
    longestSubarray(arr, k) {
        // code here
        let left = 0;
        let longest = 0;
        let map = new Map()
        let sum = 0;
        
        for(let i = 0; i < arr.length; i++){
            sum += arr[i];
            
            if(sum == k){
                longest = i + 1;
            }
            
            if(map.has(sum - k)){
                longest = Math.max(longest, i - map.get(sum - k));
            }
            
            if(!map.has(sum)){
                map.set(sum, i);
            }
        }
        return longest
    }
}
